datamodel/samples.d/generate_timestampdefinitions.py

```diff
@@ -0,0 +1,1188 @@
+    'PILO': ServiceTypeInfo('Pilotage', ['BRTH'], as_publisher_patterns(['PLT'], CARRIER_ROLES)),
+    'TOWG': ServiceTypeInfo('Towage', ['BRTH'], as_publisher_patterns(['TWG'], CARRIER_ROLES)),
+    'SHPW': ServiceTypeInfo('Shore Power', ['BRTH'], as_publisher_patterns(['SVP'], CARRIER_ROLES)),
+    'ANCO': ServiceTypeInfo('Anchorage Operations', ['ANCH'], as_publisher_patterns(['SVP'], CARRIER_ROLES)),
+    'SLUG': ServiceTypeInfo('Sludge', ['BRTH'], as_publisher_patterns(['SVP'], CARRIER_ROLES)),
+    # Services that do not have a clear-cut service name (including "null")
+    'SAFE': UNNAMED,
+    'FAST': UNNAMED,
+    'GWAY': UNNAMED,
+    'VRDY': UNNAMED,
+    NULL_VALUE: UNNAMED,  # "null" is not translated to None
+}
+REQ_PLN_ACT = frozenset({
+    'REQ',
+    'PLN',
+    'ACT',
+})
+ACT = frozenset({
+    'ACT',
+})
+EST = frozenset({
+    'EST',
+})
+EST_REQ_PLN = frozenset({
+    'EST',
+    'REQ',
+    'PLN',
+})
+EST_PLN = frozenset({
+  'EST',
+  'PLN',
+})
+EST_PLN_REQ_ACT = frozenset({
+    'EST',
+    'PLN',
+    'REQ',
+    'ACT',
+})
+VALID_JIT_VERSIONS = frozenset({
+    'jit1_0',
+    'jit1_1',
+    'jit1_2',
+})
+VALID_NEGOTIATION_CYCLES = set()
+VALID_OPERATIONS_EVENT_TYPE_CODES = set()
+VALID_PORT_CALL_PARTS = frozenset({
+    "Berth Arrival Planning",
+    "Services Planning",
+    "Pilot Boarding Place Arrival Planning And Execution, Berth Arrival Execution",
+    "Start Cargo Operations And Services",
+    "Port Departure Planning And Services Completion",
+    "Port Departure Execution",
+    "Other Services - Anchorage Planning And Execution",
+    "Other Services - Sludge Planning And Execution",
+    "Other Services - Shore Power Execution",
+    "OMIT Port Call Or Cancel A Service"
+})
+
+
+class PublisherPatternDefaultDict(defaultdict):
+
+    def __missing__(self, key):
+        # ('CA', 'TR') -> CA2TR
+        v = "2".join(key)
+        self[key] = v
+        return v
+
+
+ALL_TS = {}
+TS_PUBLISHER_PATTERN_LINKS = set()
+USED_PUBLISHER_PATTERNS = PublisherPatternDefaultDict()
+PUBLISHER_PATTERN_CA2TR = as_publisher_patterns(CARRIER_ROLES, ['TR'])
+
+# Opposite order of PUBLISHER_PATTERN_CA2TR
+PUBLISHER_PATTERN_TR2CA = [
+    PublisherPattern(publisher_role=r, primary_receiver_role=p)
+    for p, r in PUBLISHER_PATTERN_CA2TR
+]
+
+
+PUBLISHER_PATTERN_CA2ATH = as_publisher_patterns(CARRIER_ROLES, ['ATH'])
+
+
+def next_id():
+    global TS_ID_COUNTER
+    TS_ID_COUNTER += 1
+    return 'TS' + str(TS_ID_COUNTER)
+
+
+def xty_service_timestamps(
+        event_classifier_codes: Iterable[str],
+        port_call_service_type_codes: Iterable[str],
+        port_call_phase_type_codes: Iterable[str],
+        port_call_part: str,
+        provided_in_standard: str,
+        include_phase_in_name: bool = False,
+        need_vessel_position_for=frozenset(),
+        need_event_location_for=frozenset(),
+        is_cancelable: bool = True,
```

Comment:
Please add a FIXME that for `is_cancelable`  - I noted we have some `Cancel ET...` with `eventClassifierCode=EST` which is definitely wrong.  `Cancel X` events must always be `ACT`, so the code must be doing something wrong.

Reply:
If I understand you correct all Cancel timestamps must have the eventClassifierCode ACT if this is the case can it be solved with this:
`if operations_event_type_code == 'CANC':
            full_name = ''.join(('Cancel ', event_classifier_code[0], 'T<ALL>-',
                                 name_stem, phase_name_part))
            event_classifier_code = 'ACT'`

Reply:
made the fixme if the above is the correct solution then I will make the changes in a subsequent commit